File: VB/Scaffolding.ValidationErrors/DepartmentContextDataModel/UnitOfWorkSource.vb
Imports System
Imports System.Linq
Imports System.Data
Imports System.Data.Entity
Imports System.Linq.Expressions
Imports System.Collections.Generic
Imports Scaffolding.ValidationErrors.Common.Utils
Imports Scaffolding.ValidationErrors.Common.DataModel
Imports Scaffolding.ValidationErrors.Common.DataModel.EntityFramework
Imports Scaffolding.ValidationErrors.Model
Imports DevExpress.Mvvm

Namespace Scaffolding.ValidationErrors.DepartmentContextDataModel
    ''' <summary>
    ''' Provides methods to obtain the relevant IUnitOfWorkFactory.
    ''' </summary>
    Public NotInheritable Class UnitOfWorkSource

        Private Sub New()
        End Sub


        #Region "inner classes"
        Private Class DbUnitOfWorkFactory
            Implements IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork)

            Public Shared ReadOnly Instance As IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork) = New DbUnitOfWorkFactory()
            Private Sub New()
            End Sub
            Private Function IUnitOfWorkFactoryGeneric_CreateUnitOfWork() As IDepartmentContextUnitOfWork Implements IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork).CreateUnitOfWork
                Return New DepartmentContextUnitOfWork(Function() New DepartmentContext())
            End Function
        End Class

        Private Class DesignUnitOfWorkFactory
            Implements IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork)

            Public Shared ReadOnly Instance As IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork) = New DesignUnitOfWorkFactory()
            Private Sub New()
            End Sub
            Private Function IUnitOfWorkFactoryGeneric_CreateUnitOfWork() As IDepartmentContextUnitOfWork Implements IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork).CreateUnitOfWork
                Return New DepartmentContextDesignTimeUnitOfWork()
            End Function
        End Class
        #End Region

        ''' <summary>
        ''' Returns the IUnitOfWorkFactory implementation based on the current mode (run-time or design-time).
        ''' </summary>
        Public Shared Function GetUnitOfWorkFactory() As IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork)
            Return GetUnitOfWorkFactory(ViewModelBase.IsInDesignMode)
        End Function

        ''' <summary>
        ''' Returns the IUnitOfWorkFactory implementation based on the given mode (run-time or design-time).
        ''' </summary>
        ''' <param name="isInDesignTime">Used to determine which implementation of IUnitOfWorkFactory should be returned.</param>
        Public Shared Function GetUnitOfWorkFactory(ByVal isInDesignTime As Boolean) As IUnitOfWorkFactory(Of IDepartmentContextUnitOfWork)
            Return If(isInDesignTime, DesignUnitOfWorkFactory.Instance, DbUnitOfWorkFactory.Instance)
        End Function
    End Class
End Namespace

File: CS/Scaffolding.ValidationErrors/DepartmentContextDataModel/UnitOfWorkSource.cs
using System;
using System.Linq;
using System.Data;
using System.Linq.Expressions;
using System.Collections.Generic;
using Scaffolding.ValidationErrors.Common.Utils;
using Scaffolding.ValidationErrors.Common.DataModel;
using Scaffolding.ValidationErrors.Common.DataModel.DesignTime;
using Scaffolding.ValidationErrors.Common.DataModel.EntityFramework;
using Scaffolding.ValidationErrors.Model;
using DevExpress.Mvvm;
using System.Collections;
using System.ComponentModel;
using DevExpress.Data.Linq;
using DevExpress.Data.Linq.Helpers;
using DevExpress.Data.Async.Helpers;

namespace Scaffolding.ValidationErrors.DepartmentContextDataModel {
    /// <summary>
    /// Provides methods to obtain the relevant IUnitOfWorkFactory.
    /// </summary>
    public static class UnitOfWorkSource {
        /// <summary>
        /// Returns the IUnitOfWorkFactory implementation based on the current mode (run-time or design-time).
        /// </summary>
        public static IUnitOfWorkFactory<IDepartmentContextUnitOfWork> GetUnitOfWorkFactory() {
            return GetUnitOfWorkFactory(ViewModelBase.IsInDesignMode);
        }

        /// <summary>
        /// Returns the IUnitOfWorkFactory implementation based on the given mode (run-time or design-time).
        /// </summary>
        /// <param name="isInDesignTime">Used to determine which implementation of IUnitOfWorkFactory should be returned.</param>
        public static IUnitOfWorkFactory<IDepartmentContextUnitOfWork> GetUnitOfWorkFactory(bool isInDesignTime) {
            if(isInDesignTime)
                return new DesignTimeUnitOfWorkFactory<IDepartmentContextUnitOfWork>(() => new DepartmentContextDesignTimeUnitOfWork());
            return new DbUnitOfWorkFactory<IDepartmentContextUnitOfWork>(() => new DepartmentContextUnitOfWork(() => new DepartmentContext()));
        }
    }
}